examples/server/server.cpp

```diff
@@ -49,347 +119,309 @@ struct llama_server_context
         }
-      }
-    }
-    if (params.interactive && (int)embd_inp.size() <= n_consumed)
-    {
-      // check for reverse prompt
-      if (params.antiprompt.size())
-      {
-        std::string last_output;
-        for (auto id : last_n_tokens)
+        else if (mirostat == 2)
         {
-          last_output += llama_token_to_str(ctx, id);
+          static float mirostat_mu = 2.0f * mirostat_tau;
+          llama_sample_temperature(ctx, &candidates_p, temp);
+          id = llama_sample_token_mirostat_v2(ctx, &candidates_p, mirostat_tau, mirostat_eta, &mirostat_mu);
         }
-        has_next_token = true;
-        // Check if each of the reverse prompts appears at the end of the output.
-        for (std::string &antiprompt : params.antiprompt)
+        else
         {
-          if (last_output.find(antiprompt.c_str(), last_output.length() - antiprompt.length(), antiprompt.length()) != std::string::npos)
-          {
-            has_next_token = false;
-            return result;
-          }
+          // Temperature sampling
+          llama_sample_tail_free(ctx, &candidates_p, tfs_z, 1);
+          llama_sample_typical(ctx, &candidates_p, typical_p, 1);
+          llama_sample_top_p(ctx, &candidates_p, top_p, 1);
+          llama_sample_top_k(ctx, &candidates_p, top_k, 1);
+          llama_sample_temperature(ctx, &candidates_p, temp);
+          id = llama_sample_token(ctx, &candidates_p);
         }
       }
-      if (n_past > 0)
-      {
-        has_next_token = true;
-      }
+      last_n_tokens.erase(last_n_tokens.begin());
+      last_n_tokens.push_back(id);
+      num_tokens_predicted++;
     }
 
+    // add it to the context
+    embd.push_back(id);
+    result = id;
+    // decrement remaining sampling budget
+    --n_remain;
+
     if (!embd.empty() && embd.back() == llama_token_eos()) {
+        stopping_word = llama_token_to_str(ctx, embd.back());
         has_next_token = false;
+        if (verbose) {
+            fprintf(stderr, "eos token found!\n");
+        }
+        return result;
     }
 
-    if (params.interactive && n_remain <= 0 && params.n_predict != -1)
-    {
-      n_remain = params.n_predict;
-    }
-    has_next_token = n_remain != 0;
+    has_next_token = params.n_predict == -1 || n_remain != 0;
     return result;
   }
 
+  size_t findStoppingStrings(const std::string &text, const size_t last_token_size,
+                             const stop_type type)
+  {
+    size_t stop_pos = std::string::npos;
+    for (const std::string &word : params.antiprompt) {
+        size_t pos;
+        if (type == STOP_FULL) {
+            const size_t tmp = word.size() + last_token_size;
+            const size_t from_pos = text.size() > tmp ? text.size() - tmp : 0;
+            pos = text.find(word, from_pos);
+        } else {
+            pos = find_partial_stop_string(word, text);
+        }
+        if (pos != std::string::npos &&
+            (stop_pos == std::string::npos || pos < stop_pos)) {
```

Comment:
Why is `stop_pos` being compared here? It can't change after its declaration, can it?

Reply:
Ah sorry, I misread the code. I didn't see that there is a loop.